scrapy_zyte_api/providers.py

```diff
@@ -21,11 +22,29 @@ class ZyteApiProvider(PageObjectInputProvider):
 
     provided_classes = {BrowserResponse, BrowserHtml, Product}
 
+    def __init__(self, injector):
+        super().__init__(injector)
+        self._cached_instances: WeakKeyDictionary[Request, Dict] = WeakKeyDictionary()
```

Comment:
We have a similar solution in ItemProvider. I wonder if it's possible to move this cache to Injector instead, so that there is a single cache which is shared across all providers.

The first step could be using the same cached_istances storage, and keeping all cache handling code as-is. But maybe it'd be possible to improve on this further, e.g. don't pass a class to to_provide if an instance is in the cache, and also automatically add results to the cache.

Reply:
I had this in my to-do, but I wonder if we need it now that ItemProvider is gone. I am not sure in which scenario the injector a provider would get the same Request +  deps now, given duplicate request filtering.